### PR TITLE
Initial implementation of forwardPorts()

### PR DESCRIPTION
**Pull request by lmorchard**
So, here's a thing that forwards ports. @sole, @tofumatt, and/or @fwenzel - care to take a peek?

Seems to work with the half-dozen devices I hooked up to my computer - 3 android, 3 fxos.

One thing I was considering simplifying, though, is the list of ports resulting from forwardPorts(). Right now I just dumped in everything that the ADB client returned. But, it might be more handy to reduce that to just a simple list of ports - or maybe just a single port, assuming we only get one port forward to the debugging interface.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+var REMOTE_DEBUGGER_SOCKET = 'localfilesystem:/data/local/debugger-socket';
+
+module.exports = forwardPorts;
+
+function forwardPorts(options) {
+  options = options || {};
+  var devices = options.devices;
+
+  var client = adb.createClient();
```
Comment (sole) [inline index.js]: I'd suggest renaming this variable to `adbClient` or similar. It confused me with _our_ `firefoxClient` and I was about to write a comment saying that you should not tie forwards to clients (firefox clients) blablabla

File: index.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    });
+
+  });
+}
+
+function forwardDevices(client, devices) {
+  return Promise.all(devices.map(function(device) {
+
+    if (device.ports.length > 0) {
+      // Skip forwarding for this device, but ensure TCPIP ports are in a more
```
Comment (sole) [inline index.js]: maybe add an example here of initial format, end format? so I don't need to mentally parse the string manipulations and can 'get it' without even thinking

File: package.json [hunk 3]
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "node-firefox-forward-ports",
+  "version": "1.0.0",
+  "description": "Find attached FirefoxOS devices using ADB",
```
Comment (sole) [inline package.json]: dear native English speakers, is _attached_ the right word? would _connected_ or _plugged in_ be better? I feel like an attachment is more superficial?

also, we're not finding them we're forwarding them

File: package.json [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  "main": "index.js",
+  "dependencies": {
+    "adbkit": "^2.1.6",
+    "es6-promise": "^2.0.1",
+    "portfinder": "^0.4.0"
+  },
+  "devDependencies": {
+    "gulp": "^3.8.10",
+    "gulp-jscs": "^1.3.1",
```
Comment (sole) [inline package.json]: do you _actually_ need all these gulp-\* dependencies or are them copy paste remains? IIRC they are included by the `build-tools` module and you should not need to install them in the top module

File: tests/unit/test.index.js [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      test.ok(false);
+      test.done();
+    }).catch(function(err) {
+      test.done();
+    });
+  },
+
+  'forwardPorts() should yield an error if the devices option is not an array': function(test) {
+    forwardPorts({
+      devices: 'LOL NOT AN ARRAY'
```
Comment (sole) [inline tests/unit/test.index.js:26]: ppfffttt lolol

File: README.md [hunk 6]
```diff
@@ -1,2 +1,96 @@
...
+npm install
+```
+
+### npm
+
+```bash
+npm install node-firefox-forward-ports
+```
+
+## Usage
```
Comment (sole) [inline README.md:38]: Usage should show the _usage_ of the module without other things. I.e. show me how does the API look like, don't make me go and look at the code to see what kind of parameters are you accepting

Comment (sole) [inline README.md:38]: I think this header should be renamed to "Example" and then you could add the Usage section above :)

File: README.md [hunk 7]
```diff
@@ -1,2 +1,96 @@
...
+
+findDevices().then(function (devices) {
+  return forwardPorts({
+    devices: devices
+  });
+}).then(function(results) {
+  console.log(results);
+});
+
+[
```
Comment (sole) [inline README.md:92]: add a "would return something like this..." line or it's like what? where does this come from? don't confuse people!

File: README.md [hunk 8]
```diff
@@ -1,2 +1,96 @@
...
+  {
+    "id": "3739ced5",
+    "type": "device",
+    "isFirefoxOS": true,
+    "ports": [
```
Comment (tofumatt) [inline README.md:97]: Will this return more than one set of open ports per-device? If not this info doesn't look like it should be nested.

File: index.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+      // friendly format.
+      device.ports.forEach(function(port) {
+        port.port = (port.local.indexOf('tcp:') === -1) ? null :
+          port.local.replace('tcp:', '');
```
Comment (tofumatt) [inline index.js]: I found the indentation of this a bit hard to follow. Would:

``` javascript
        port.port = (port.local.indexOf('tcp:') === -1) ? null :
                    port.local.replace('tcp:', '');
```

be better for you? I think I can follow it easier.

File: package.json (hunk 3 above)
Comment (tofumatt) [inline package.json]: I would say "forward connected Firefox OS devices using ADB".

File: package.json (hunk 4 above)
Comment (tofumatt) [inline package.json]: I believe this is the case, and because `build-tools` uses `gulp` as an actual dependency you don't need to list it here. I think.

**Comment by tofumatt** [comment]
Mostly good to me, @sole covered documentation stuff well already, I just had a thought on the format of the returned JSON from the API.

File: index.js (hunk 1 above)
Comment (lmorchard) [inline index.js]: Good point - I'll submit a PR to find-devices too, since I just named it `client` there, too.

File: package.json (hunk 3 above)
Comment (lmorchard) [inline package.json]: Huh, interesting: I see connected & attached as synonyms for wired devices. But, if it reads differently to you, I'll change it. 

I used this wording in the README - "Forward local ports to remote debugging interfaces on connected Firefox OS devices"

File: package.json (hunk 4 above)
Comment (lmorchard) [inline package.json]: Yup, this is copypasta from another node-firefox module. Most of them have all this stuff in `devDependencies`.

Removing all except `gulp` and reinstalling seems to work fine. Might be worth circling back to the other modules and doing the same

File: README.md (hunk 8 above)
Comment (lmorchard) [inline README.md:97]: Yes, it can return multiple ports per device. Though I'm filtering out everything except debugging sockets, you can have many local ports forwarded to that socket - e.g. you manually ran `adb forward` at some point, and the ADB Helper in Firefox later forwarded another one.

Comment (lmorchard) [inline README.md:97]: For example, this is what I get after I manually forward two ports to the same socket:

``` json
 {
  "id": "3739ced5",
  "type": "device",
  "isFirefoxOS": true,
  "ports": [
   {
    "serial": "3739ced5",
    "local": "tcp:8002",
    "remote": "localfilesystem:/data/local/debugger-socket",
    "port": "8002"
   },
   {
    "serial": "3739ced5",
    "local": "tcp:8003",
    "remote": "localfilesystem:/data/local/debugger-socket",
    "port": "8003"
   }
  ]
 }
```

Thinking I should at least delete the `serial` property, because it's redundant. The `port` will be important for Firefox client connections.

The `remote` property gets important if we want to expand this to, for example, forward [Marionette test ports](https://developer.mozilla.org/en-US/docs/Marionette_Test_Runner). Then, those will have `"remote": "tcp:2828"`, and we'll want to use that to filter & select appropriate port forwards.

File: README.md (hunk 7 above)
Comment (lmorchard) [inline README.md:92]: Adding this!

File: README.md (hunk 6 above)
Comment (lmorchard) [inline README.md:38]: Yup, good idea. Doing some tweaking here, hopefully it will read better.

File: index.js (hunk 9 above)
Comment (lmorchard) [inline index.js]: Yeah, this whole assignment is awkward. Let's see if that indentation improves it!

File: README.md (hunk 8 above)
Comment (tofumatt) [inline README.md:97]: Cool, sounds good, and makes sense to me. I didn’t quite follow that at first so it might be worth explaining somewhere like you just did, but I get it now :)

File: index.js (hunk 9 above)
Comment (tofumatt) [inline index.js]: It also might be a case of “using an if statement is more lines and repetition but improves readability”. I actually love the ternary operator but it might not improve legibility here.

File: README.md (hunk 8 above)
Comment (lmorchard) [inline README.md:97]: I just added some notes on this structure in the README example

File: README.md [hunk 10]
```diff
@@ -1,2 +1,135 @@
...
+
+## Usage
+
+```javascript
+var forwardPorts = require('node-firefox-forward-ports');
+
+// Forwards ports as necessary, while reusing existing ports.
+// Returns the list of devices annotated with details on forwarded ports.
+forwardPorts({
+  devices: [
```
Comment (lmorchard) [inline README.md:58]: I'm also tempted to simplify the API from `forwardPorts({ devices: devices })` to just `forwardPorts(devices)`.

Then, you could do something like `findDevices().then(forwardPorts).then(...)`. But, that would make it hard / awkward to add additional options in the future.

Comment (tofumatt) [inline README.md:58]: I actually like that idea.

And `devices` is an array, right? No reason you couldn't support both APIs; if it's an array then treat it like the new API, if it's a key/value set use the current API.

I propose that!

**Comment by lmorchard** [comment]
Just added another commit to support a simplified API. That allows a quick device detection and port forwarding promise chain that looks like so:

``` javascript
findDevices().then(forwardPorts).then(console.log)
```

**Comment by tofumatt** [comment]
:+1: Yay, awesome!
